core/new-gui/src/app/workspace/service/drag-drop/drag-drop.service.ts

```diff
@@ -0,0 +1,214 @@
+    this.elementOperatorTypeMap.set(dragElementID, operatorType);
+
+    // register callback functions for jquery UI
+    jQuery('#' + dragElementID).draggable({
+      helper: () => this.createFlyingOperatorElement(operatorType),
+      // declare event as type any because the jQueryUI type declaration is wrong
+      // it should be of type JQuery.Event, which is incompatible with the the declared type Event
+      start: (event: any, ui) => this.handleOperatorStartDrag(event, ui)
+    });
+  }
+
+  /**
+   * This function should be only used by the Workflow Editor Componenet
+   *  to register itself as a droppable area.
+  */
+  public registerWorkflowEditorDrop(dropElementID: string): void {
+    jQuery('#' + dropElementID).droppable({
+      drop: (event: any, ui) => this.handleOperatorDrop(event, ui)
+    });
+  }
+
+  /**
+   * Creates a DOM Element that visually looks identical to the operator when dropped on main workflow editor
+   *
+   * This function temporarily creates a DOM element which contains a JointJS paper that has the exact size of the operator,
+   *    then create the operator Element based on the operatorType and make it fully occupy the JointJS paper.
+   *
+   * The temporary JointJS paper element has ID "flyingJointPaper". This DOM elememtn will be destroyed by jQueryUI when the dragging ends.
+   *
+   * @param operatorType - the type of the operator
+   */
+  private createFlyingOperatorElement(operatorType: string): JQuery<HTMLElement> {
+    this.currentOperatorType = operatorType;
+
+    // create a temporary ghost element
+    jQuery('body').append('<div id="flyingJointPaper" style="position:fixed;z-index:100;pointer-event:none;"></div>');
+
+    // create an operator and get the UI element from the operator type
+    const operator = this.workflowUtilService.getNewOperatorPredicate(operatorType);
+    const operatorUIElement = this.jointUIService.getJointOperatorElement(operator, { x: 0, y: 0 });
+
+    // create the jointjs model and paper of the ghost element
+    const tempGhostModel = new joint.dia.Graph();
+    const tempGhostPaper = new joint.dia.Paper({
+      el: jQuery('#flyingJointPaper'),
+      width: JointUIService.DEFAULT_OPERATOR_WIDTH,
+      height: JointUIService.DEFAULT_OPERATOR_HEIGHT,
+      model: tempGhostModel,
+    });
+
+    // add the operator JointJS element to the paper
+    tempGhostModel.addCell(operatorUIElement);
+
+    // return the jQuery object of the DOM Element
+    return jQuery('#flyingJointPaper');
+  }
+
+  /**
+   * Hanlder function for jQueryUI's drag started event.
+   * It converts the event to the drag started Subject.
+   *
+   * @param event JQuery.Event type, although JQueryUI typing says the type is Event, the object's actual type is JQuery.Event
+   * @param ui jQueryUI Draggable Event UI
+   */
+  private handleOperatorStartDrag(event: JQuery.Event, ui: JQueryUI.DraggableEventUIParams): void {
+    const eventElement = event.toElement;
```

Comment:
eventElement type not mentioned. How is this compiling? Similarly, below operatorType's type isn't specified.

Reply:
no need to declare type when declaring in a variable, ts will do its type inference

https://basarat.gitbooks.io/typescript/docs/types/type-inference.html

Reply:
there are also cases that I explicitly declare the variable type although I don't need to:

for example, 
the return type signature of `jointGraph.getCell()` is `joint.dia.Cell`, however, that definition is wrong because I know it returns `undefined` if getting a cell with ID that doesn't exist. So I write 
`const cell: joint.dia.Cell | undefined = jointGraph.getCell(operatorID)`
to let Typescript remind me to do property error handling.

Another example is when I sometimes declare a class variable, for example
`private a: boolean = true;`
simply to make it more easy to read, because this variable might be used in a lot of places. That is less of a concern for local variables inside functions.